actors/market/src/lib.rs

```diff
@@ -355,61 +386,39 @@ impl Actor {
-                let params = datacap_transfer_request(
-                    &Address::new_id(client_id),
-                    vec![alloc_request_for_deal(&deal, rt.policy(), curr_epoch)],
-                )?;
-                let alloc_ids = rt
-                    .send(
-                        &DATACAP_TOKEN_ACTOR_ADDR,
-                        ext::datacap::TRANSFER_FROM_METHOD as u64,
-                        serialize(&params, "transfer parameters")?,
-                        TokenAmount::zero(),
-                    )
-                    .and_then(|ret| datacap_transfer_response(&ret));
-                match alloc_ids {
-                    Ok(ids) => {
-                        if ids.len() != 1 {
-                            return Err(actor_error!(
-                                unspecified,
-                                "expected 1 allocation ID, got {:?}",
-                                ids
-                            ));
-                        }
-                        ids[0]
-                    }
-                    Err(e) => {
-                        info!(
-                            "invalid deal {}: failed to allocate datacap for verified deal: {}",
-                            di, e
-                        );
-                        continue;
-                    }
-                }
-            } else {
-                NO_ALLOCATION_ID
-            };
-
             total_provider_lockup = provider_lockup;
             total_client_lockup.insert(client_id, client_lockup);
             proposal_cid_lookup.insert(pcid);
-            valid_deals.push(ValidDeal {
-                proposal: deal.proposal,
-                cid: pcid,
-                allocation: allocation_id,
-            });
+            valid_deals.push(ValidDeal { proposal: deal.proposal, cid: pcid });
             valid_input_bf.set(di as u64)
         }
+        let mut alloc_reqs: BTreeMap<ActorID, Vec<AllocationRequest>> = BTreeMap::new();
+        for valid_deal in valid_deals.iter() {
+            alloc_reqs
+                .entry(valid_deal.proposal.client.id().unwrap())
+                .or_default()
+                .push(alloc_request_for_deal(&valid_deal.proposal, rt.policy(), curr_epoch));
+        }
+
+        let mut allocation_map: BTreeMap<ActorID, Vec<AllocationID>> = BTreeMap::new();
+        for client_datacap in all_client_datacap.iter() {
+            let params = datacap_transfer_request(
+                &Address::new_id(*client_datacap.0 as u64),
+                alloc_reqs.get(client_datacap.0).unwrap().clone(),
+            )?;
+            let alloc_ids = rt
+                .send(
+                    &DATACAP_TOKEN_ACTOR_ADDR,
+                    ext::datacap::TRANSFER_FROM_METHOD as u64,
+                    serialize(&params, "transfer parameters")?,
+                    TokenAmount::zero(),
+                )
+                .and_then(|ret| datacap_transfer_response(&ret));
```

Comment:
What happens if this fails? Don't unwrap below, we need to handle it properly. It's unexpected, for sure. I guess you need to abort in order to have the property below that any verified, valid deal has an allocation ID, but let's do so with an explicit check and message describing the error.